### PR TITLE
Change method dispatch

### PR DESCRIPTION
**Pull request by kdheepak**
- HydroGen
- ElectricLoads
- RenewableGen
- Storage
- HydroPumpedStorage
- FlowActivePowerVariable

### CODE REVIEW

File: src/devices_models/devices/storage.jl [hunk 1]
```diff
@@ -1,66 +1,43 @@
...
+############## ReactivePowerVariable, Storage ####################
+
+get_variable_binary(::ReactivePowerVariable, ::Type{<:PSY.Storage}) = false
+
+get_variable_expression_name(::ReactivePowerVariable, ::Type{<:PSY.Storage}) = :nodal_balance_reactive
+
+############## EnergyVariable, Storage ####################
+
+get_variable_binary(::EnergyVariable, ::Type{<:PSY.Storage}) = false
+# TODO: why isn't this defined
```
Comment (daniel-thom) [inline src/devices_models/devices/storage.jl:32]: Do these need to be implemented before merging?

File: src/PowerSimulations.jl [hunk 2]
```diff
@@ -1,6 +1,8 @@
 isdefined(Base, :__precompile__) && __precompile__()
 module PowerSimulations
 
+using InteractiveUtils
```
Comment (jd-lara) [inline src/PowerSimulations.jl:4]: why using here and not Import? is this a left over from other stuff?

File: src/network_models/network_constructor.jl [hunk 3]
```diff
@@ -54,7 +54,7 @@ function construct_network!(
             btype,
             Vector([[b for b in dc_branches if typeof(b) == btype]]),
         )
-        flow_variables!(psi_container, StandardPTDFModel, typed_dc_branches)
+        add_variables!(psi_container, StandardPTDFModel(), typed_dc_branches)
```
Comment (daniel-thom) [inline src/network_models/network_constructor.jl:57]: Why did you change this to an instance instead of a type?

File: src/core/psi_container.jl [hunk 4]
```diff
@@ -17,7 +17,7 @@ mutable struct PSIContainer
         settings::PSISettings,
         jump_model::Union{Nothing, JuMP.AbstractModel},
     )
-        #PSY.check_forecast_consistency(sys)
+        # PSY.check_forecast_consistency(sys)
```
Comment (jd-lara) [inline src/core/psi_container.jl:20]: We can remove this line, this is no longer a PSY function.

**Review by daniel-thom — COMMENTED** [review]
Overall this is exactly like what you proposed and looks good. Just a few minor comments.

File: src/devices_models/devices/DC_branches.jl [hunk 5]
```diff
@@ -14,9 +14,9 @@ flow_variables!(
     devices::IS.FlattenIteratorWrapper{<:PSY.DCBranch},
 ) = nothing
 
-function flow_variables!(
+function add_variables!(
     psi_container::PSIContainer,
-    system_formulation::Type{StandardPTDFModel},
+    system_formulation::StandardPTDFModel,
```
Comment (jd-lara) [inline src/devices_models/devices/DC_branches.jl:19]: The system formulation variable is not used here. It can be removed.

File: src/devices_models/devices/common/add_variable.jl [hunk 6]
```diff
@@ -52,7 +6,11 @@ function add_variables!(
     ::Type{T},
     devices::Union{Vector{U}, IS.FlattenIteratorWrapper{U}},
 ) where {T <: VariableType, U <: PSY.Component}
-    _add_variables!(psi_container, devices, AddVariableSpec(T, U, psi_container))
+    add_variable!(
+        psi_container,
+        T(),
```
Comment (jd-lara) [inline src/devices_models/devices/common/add_variable.jl:11]: Is it better to evaluate here instead of passing the type? I don't understand the reason for doing that instead of passing `T`. The code style in PSI has used T as type so far.

File: src/devices_models/devices/common/add_variable.jl [hunk 7]
```diff
@@ -62,47 +20,13 @@ function add_variables!(
...
-    else
-        filter_func = spec.devices_filter_func
-    end
-    if !isnothing(filter_func)
-        devices = filter!(filter_func, collect(devices))
-    end
-
     add_variable!(
         psi_container,
+        T(),
```
Comment (jd-lara) [inline src/devices_models/devices/common/add_variable.jl:27]: same here

File: src/devices_models/devices/common/range_constraint.jl [hunk 8]
```diff
@@ -150,7 +150,7 @@ function device_semicontinuousrange!(
     return
 end
 
-#This function looks suspicious and repetitive. Needs verification
+# This function looks suspicious and repetitive. Needs verification
```
Comment (jd-lara) [inline src/devices_models/devices/common/range_constraint.jl:153]: Is this comment still valid?

File: src/devices_models/devices/electric_loads.jl [hunk 9]
```diff
@@ -4,44 +4,31 @@ struct StaticPowerLoad <: AbstractLoadFormulation end
...
-    return AddVariableSpec(;
-        variable_name = make_variable_name(T, U),
-        binary = false,
-        expression_name = :nodal_balance_reactive,
-        sign = -1.0,
-        lb_value_func = x -> 0.0,
-        ub_value_func = x -> PSY.get_max_reactive_power(x),
-    )
-end
+get_variable_sign(_, ::Type{<:PSY.ElectricLoad}) = -1.0
```
Comment (jd-lara) [inline src/devices_models/devices/electric_loads.jl:9]: we have been using empty spaces for this. If we are moving into this style is important to document it and make it consistent.

File: src/devices_models/devices/electric_loads.jl [hunk 10]
```diff
@@ -4,44 +4,31 @@ struct StaticPowerLoad <: AbstractLoadFormulation end
...
-    ::Type{U},
-    ::PSIContainer,
-) where {T <: OnVariable, U <: PSY.ElectricLoad}
-    return AddVariableSpec(; variable_name = make_variable_name(T, U), binary = true)
-end
+########################### ActivePowerVariable, ElectricLoad ####################################
+
+get_variable_binary(::ActivePowerVariable, ::Type{<:PSY.ElectricLoad}) = false
+
+get_variable_expression_name(::ActivePowerVariable, ::Type{<:PSY.ElectricLoad}) = :nodal_balance_active
```
Comment (jd-lara) [inline src/devices_models/devices/electric_loads.jl:15]: This part could be problematic in the future as a variable could be added to different expressions. I think we talked about it before. For now is ok as nothing is breaking but this could require making It a function of the formulation too.

File: src/network_models/network_constructor.jl (hunk 3 above)
Comment (jd-lara) [inline src/network_models/network_constructor.jl:57]: Same comment as above.

**Review by jd-lara — COMMENTED** [review]
Overall looks great, I have some questions about the changes made. Mostly there are some patterns that aren't consistent with the rest of the code base that I think we should make consistent or avoid.

File: src/core/psi_container.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (kdheepak) [reply src/core/psi_container.jl]: Resolved in c359a46a.

File: src/devices_models/devices/common/add_variable.jl [hunk 11]
```diff
@@ -222,3 +205,44 @@ function set_variable_bounds!(
...
+    else
+        initial_value = nothing
+    end
+
+    add_variable!(
+        psi_container,
+        OnVariable(),
+        devices,
+    )
+    varstatus = get_variable(psi_container, OnVariable, PSY.ThermalMultiStart)
```
[the remark this answers is not in the document]
Comment (kdheepak) [reply src/devices_models/devices/common/add_variable.jl]: Resolved in c359a46a.

File: src/devices_models/devices/storage.jl (hunk 1 above)
Comment (kdheepak) [reply src/devices_models/devices/storage.jl]: I discussed this with @jd-lara and he said it wasn't required. I forgot to remove the comments.

Resolved in c359a46a.

File: src/PowerSimulations.jl (hunk 2 above)
Comment (kdheepak) [reply src/PowerSimulations.jl]: This is a leftover from a debugging session. Resolved in f877cf4d.

File: src/core/psi_container.jl (hunk 4 above)
Comment (kdheepak) [reply src/core/psi_container.jl]: Resolved in c359a46.

File: src/devices_models/devices/common/add_variable.jl (hunk 6 above)
[the remark this answers is not in the document]
Comment (kdheepak) [reply src/devices_models/devices/common/add_variable.jl:11]: @jd-lara has made similar comments, so I'm addressing them all here.

In this PR we are doing the following:

```julia
get_variable_lower_bound(::ActivePowerVariable, d::PSY.ThermalGen, _) = PSY.get_active_power_limits(d).min
get_variable_upper_bound(::ActivePowerVariable, d::PSY.ThermalGen, _) = PSY.get_active_power_limits(d).max
```

If we passed just the type, these methods would become this instead:

```julia
get_variable_lower_bound(::Type{<:ActivePowerVariable}, d::PSY.ThermalGen, _) = PSY.get_active_power_limits(d).min
get_variable_upper_bound(::Type{<:ActivePowerVariable}, d::PSY.ThermalGen, _) = PSY.get_active_power_limits(d).max
```

I think this is more complicated to read and understand for the average user.

Also, previously, we were always passing the types and the types alone when creating a "VariableSpec" (i.e. we were using `::Type{<:ThermalGen}` and returning anonymous functions that accepted an arugment to `d::PSY.ThermalGen`). 

Now however, we are using an instance of the object and hence have to define `d::PSY.ThermalGen`. Defining the other argument as `ActivePowerVariable` felt more natural to me.

I think this change to using instances in cases where the compiler can make optimizations makes the code easier to read, easier to write, and has no cost to performance.

Happy to discuss this further in a call.

File: src/devices_models/devices/common/add_variable.jl [hunk 12]
```diff
@@ -141,25 +65,79 @@ If binary = true:
...
+    for t in time_steps, d in devices
+        name = PSY.get_name(d)
+        variable[name, t] = JuMP.@variable(
+            psi_container.JuMPmodel,
+            base_name = "$(var_name)_{$(name), $(t)}",
+            binary = binary
+        )
+
+        ub = get_variable_upper_bound(variable_type, d, psi_container.settings)
+        !isnothing(ub) && JuMP.set_upper_bound(variable[name, t], ub)
```
[the remark this answers is not in the document]
Comment (kdheepak) [reply src/devices_models/devices/common/add_variable.jl:96]: Makes sense. Would you like me to make these changes in this PR? Current status of the codebase is as follows:

```plaintext
$ rg isnothing src --count-matches | wc -l
23

$ rg '=== nothing' src --count-matches | wc -l
2

$ rg '!== nothing' src --count-matches | wc -l
0
```

[the remark this answers is not in the document]
Comment (daniel-thom) [reply src/devices_models/devices/common/add_variable.jl:96]: I personally started writing `===` instead of `isnothing` but didn't sweep through the codebases and change all occurrences. I also don't know if Jose came to the same conclusion.

File: src/devices_models/devices/common/add_variable.jl (hunk 6 above)
[the remark this answers is not in the document]
Comment (daniel-thom) [reply src/devices_models/devices/common/add_variable.jl:11]: I was mostly just asking why you made the change. Given that rationale, why didn't you change `get_variable_binary`, `get_variable_expression_name`, and `get_variable_sign` to accept instances of PSY.Component? Instances are available at the call sites.

As Jose said, PSI does heavily use a paradigm of passing the type to functions where no instance would ever be used.

[the remark this answers is not in the document]
Comment (kdheepak) [reply src/devices_models/devices/common/add_variable.jl:11]: > Given that rationale, why didn't you change get_variable_binary, get_variable_expression_name, and get_variable_sign to accept instances of PSY.Component? Instances are available at the call sites.

That would be clearer. If @jd-lara is on board, I'll make that change.

File: src/devices_models/devices/DC_branches.jl (hunk 5 above)
Comment (kdheepak) [reply src/devices_models/devices/DC_branches.jl]: Resolved in 40e2de21.

File: src/devices_models/devices/common/range_constraint.jl (hunk 8 above)
Comment (kdheepak) [reply src/devices_models/devices/common/range_constraint.jl:153]: Probably not as originally intended but when I refactor range constraints I'll remove this comment. This will be resolved in a future PR.

File: src/devices_models/devices/electric_loads.jl (hunk 9 above)
Comment (kdheepak) [reply src/devices_models/devices/electric_loads.jl:9]: Can you clarify what you've been using empty spaces for?

File: src/devices_models/devices/electric_loads.jl (hunk 10 above)
Comment (kdheepak) [reply src/devices_models/devices/electric_loads.jl:15]: We can extend it to support future use cases.

File: src/devices_models/devices/DC_branches.jl (hunk 5 above)
Comment (kdheepak) [reply src/devices_models/devices/DC_branches.jl]: Resolved in 461c2777 instead.

File: src/devices_models/devices/common/add_variable.jl (hunk 6 above)
[the remark this answers is not in the document]
Comment (jd-lara) [reply src/devices_models/devices/common/add_variable.jl:11]: I prefer we keep the current paradigm. Most of the collaborators and developers are already used to this. It is also consistent with the interfaces in PSY. For instance, get_component() uses a type as input. Most of our current users are already used to passing types and not instances.

File: src/devices_models/devices/common/add_variable.jl (hunk 12 above)
[the remark this answers is not in the document]
Comment (jd-lara) [reply src/devices_models/devices/common/add_variable.jl:96]: making the change in this PR is appropriate.

File: src/devices_models/devices/electric_loads.jl (hunk 9 above)
Comment (jd-lara) [reply src/devices_models/devices/electric_loads.jl:9]: when the argument isn't used in the specific function for instance 

```julia
function foo(val::Int, other_val::Int)
return val + other_val
end

function foo(val::Int, ::Any)
return val
end
```

File: src/devices_models/devices/DC_branches.jl [hunk 13]
```diff
@@ -127,7 +127,7 @@ function branch_rate_constraints!(
     psi_container::PSIContainer,
     devices::IS.FlattenIteratorWrapper{B},
     model::DeviceModel{B, <:AbstractDCLineFormulation},
-    system_formulation::Union{
+    ::Union{
```
Comment (jd-lara) [inline src/devices_models/devices/DC_branches.jl:130]: This union is redundant (I know you didn't introduce it). 

since 

```julia 
PM.AbstractActivePowerModel <: PM.AbstractPowerModel
```

File: src/devices_models/devices/hydro_generation.jl [hunk 14]
```diff
@@ -11,164 +11,95 @@ struct HydroCommitmentRunOfRiver <: AbstractHydroUnitCommitment end
...
-        expression_name = :nodal_balance_reactive,
-        initial_value_func = x -> PSY.get_reactive_power(x),
-        lb_value_func = x -> PSY.get_reactive_power_limits(x).min,
-        ub_value_func = x -> PSY.get_reactive_power_limits(x).max,
-    )
-end
+get_variable_initial_value(pv::ActivePowerVariable, d::PSY.HydroGen, settings) =
+    get_variable_initial_value(pv, d, get_warm_start(settings) ? WarmStartVariable() : ColdStartVariable())
+get_variable_initial_value(::ActivePowerVariable, d::PSY.HydroGen, ::WarmStartVariable) = PSY.get_active_power(d)
+get_variable_initial_value(::ActivePowerVariable, d::PSY.HydroGen, ::ColdStartVariable) = nothing
```
Comment (jd-lara) [inline src/devices_models/devices/hydro_generation.jl:22]: we don't have coldstart for Hydro. This is only for thermals.

File: src/devices_models/devices/hydro_generation.jl [hunk 15]
```diff
@@ -11,164 +11,95 @@ struct HydroCommitmentRunOfRiver <: AbstractHydroUnitCommitment end
...
-        variable_name = make_variable_name(T, U),
-        binary = false,
-        expression_name = :nodal_balance_reactive,
-        initial_value_func = x -> PSY.get_reactive_power(x),
-        lb_value_func = x -> PSY.get_reactive_power_limits(x).min,
-        ub_value_func = x -> PSY.get_reactive_power_limits(x).max,
-    )
-end
+get_variable_initial_value(pv::ActivePowerVariable, d::PSY.HydroGen, settings) =
+    get_variable_initial_value(pv, d, get_warm_start(settings) ? WarmStartVariable() : ColdStartVariable())
```
Comment (jd-lara) [inline src/devices_models/devices/hydro_generation.jl:20]: same here

File: src/devices_models/devices/hydro_generation.jl [hunk 16]
```diff
@@ -11,164 +11,95 @@ struct HydroCommitmentRunOfRiver <: AbstractHydroUnitCommitment end
...
-) where {T <: EnergyVariable, U <: PSY.HydroGen}
-    return AddVariableSpec(;
-        variable_name = make_variable_name(T, U),
-        binary = false,
-        initial_value_func = x -> PSY.get_initial_storage(x),
-        lb_value_func = x -> 0.0,
-        ub_value_func = x -> PSY.get_storage_capacity(x),
-    )
-end
+get_variable_lower_bound(::ActivePowerVariable, d::PSY.HydroGen, _) = PSY.get_active_power_limits(d).min
```
Comment (jd-lara) [inline src/devices_models/devices/hydro_generation.jl:24]: Can you make another function here that dispatches on the formulation `HydroDispatchRunOfRiver` to return 0.0. This is related to #563.

File: src/devices_models/devices/common/add_variable.jl (hunk 12 above)
[the remark this answers is not in the document]
Comment (kdheepak) [reply src/devices_models/devices/common/add_variable.jl]: Resolved in https://github.com/NREL-SIIP/PowerSimulations.jl/pull/560/commits/7fb5eda67b7873f612ac8be80e7f358bea57c2ca.

File: src/devices_models/devices/common/add_variable.jl (hunk 6 above)
[the remark this answers is not in the document]
Comment (kdheepak) [reply src/devices_models/devices/common/add_variable.jl:11]: After having discussions over the phone and chat, we decided that we should make the type / instance changes in this case. The current code is backward compatible and will not affect users. We also decided that we would approach future type / instance changes on a case by case basis.

Some references:

https://docs.julialang.org/en/v1/manual/style-guide/#Avoid-confusion-about-whether-something-is-an-instance-or-a-type

https://docs.julialang.org/en/v1/manual/style-guide/#Don't-use-unnecessary-static-parameters

File: src/devices_models/devices/electric_loads.jl (hunk 9 above)
Comment (kdheepak) [reply src/devices_models/devices/electric_loads.jl:9]: I believe we discussed that using `_` is less noisy that using `::Any`. Can someone refresh my memory?

File: src/devices_models/devices/hydro_generation.jl (hunk 14 above)
Comment (kdheepak) [reply src/devices_models/devices/hydro_generation.jl]: Resolved in https://github.com/NREL-SIIP/PowerSimulations.jl/pull/560/commits/20ab12eff7f5adf899419a65ff2d2587a4cb9a95

File: src/devices_models/devices/DC_branches.jl (hunk 13 above)
Comment (kdheepak) [reply src/devices_models/devices/DC_branches.jl]: Resolved in https://github.com/NREL-SIIP/PowerSimulations.jl/pull/560/commits/20ab12eff7f5adf899419a65ff2d2587a4cb9a95

File: src/devices_models/devices/hydro_generation.jl (hunk 15 above)
Comment (kdheepak) [reply src/devices_models/devices/hydro_generation.jl]: Resolved in https://github.com/NREL-SIIP/PowerSimulations.jl/pull/560/commits/20ab12eff7f5adf899419a65ff2d2587a4cb9a95.

File: src/devices_models/devices/hydro_generation.jl (hunk 16 above)
Comment (kdheepak) [reply src/devices_models/devices/hydro_generation.jl:22]: Resolved in https://github.com/NREL-SIIP/PowerSimulations.jl/pull/560/commits/28d99027f2bd2e4140806e0cdb92ebfed64f50cf.

File: src/devices_models/devices/electric_loads.jl (hunk 9 above)
Comment (daniel-thom) [reply src/devices_models/devices/electric_loads.jl:9]: I'm not sure that we reached consensus. I have no idea what Julia people would say. Here are my fairly weak opinions. I would go along with anything.
- In general passing arguments to functions that don't get used can be indicative of a sub-optimal design. In the cases here I think it makes sense. Hopefully, we don't have this pattern in many places.
- If it is a one-line function then `_` is less noisy.
- If it is a multi-line function then including the name of the variable at least informs the reader of what is being passed (to match the interface).

```
foo(val::Int, _) = val

function foo(val::Int, other_val)
    return val
end
```

Comment (jd-lara) [reply src/devices_models/devices/electric_loads.jl:11]: Just make sure that all the code base is consistent. I prefer not to use _ and pass the expected type.

**Review by daniel-thom — APPROVED** [review]
Looks great. Very nice improvement.

File: src/devices_models/devices/electric_loads.jl (hunk 9 above)
Comment (kdheepak) [reply src/devices_models/devices/electric_loads.jl:11]: > If it is a one-line function then _ is less noisy.

I agree with this. Let me made this change for just one line functions.

Comment (kdheepak) [reply src/devices_models/devices/electric_loads.jl:11]: Okay, it is only implemented for one line functions.

**Review by jd-lara — APPROVED** [review]
There are some areas where this code will have some conflicts with the simulation store code. Mostly over white space changes.

File: src/devices_models/devices/electric_loads.jl (hunk 9 above)
Comment (kdheepak) [reply src/devices_models/devices/electric_loads.jl:11]: > In general passing arguments to functions that don't get used can be indicative of a sub-optimal design.

These methods are specialized. Here's two examples:

Call site:

```julia
lb = get_variable_upper_bound(variable_type, d, psi_container.settings)
...
init = get_variable_initial_value(variable_type, d, psi_container.settings)
```

For `get_variable_initial_value` the function is defined like this:

```julia
get_variable_initial_value(pv::ActivePowerVariable, d::PSY.ThermalMultiStart, settings) =
    get_variable_initial_value(pv, d, get_warm_start(settings) ? WarmStartVariable() : ColdStartVariable())
get_variable_initial_value(::ActivePowerVariable, d::PSY.ThermalMultiStart, ::WarmStartVariable) = PSY.get_active_power(d)
get_variable_initial_value(::ActivePowerVariable, d::PSY.ThermalMultiStart, ::ColdStartVariable) = nothing
```

For `get_variable_upper_bound` the function is defined like this:

```julia
get_variable_upper_bound(::ActivePowerVariable, d::PSY.ThermalMultiStart, _) = PSY.get_active_power_limits(d).max
```

My thinking here is that:

1) `settings` is untyped in `get_variable_initial_value` so it should be untyped in `get_variable_upper_bound`. 
2) `::WarmStartVariable` / `::ColdStartVariable` is used for dispatch in `get_variable_initial_value`. We leave out the variable name here because it doesn't matter, and I'm doing the same for `get_variable_upper_bound`.

File: src/devices_models/devices/hydro_generation.jl (hunk 16 above)
Comment (kdheepak) [reply src/devices_models/devices/hydro_generation.jl:24]: Can you clarify exactly what needs to be added here?

Comment (jd-lara) [reply src/devices_models/devices/hydro_generation.jl:24]: add 

```julia
get_variable_lower_bound(::ActivePowerVariable, d::PSY.HydroGen, ::HydroDispatchRunOfRiver) = 0.0
```

Comment (kdheepak) [reply src/devices_models/devices/hydro_generation.jl:24]: Resolved in https://github.com/NREL-SIIP/PowerSimulations.jl/pull/560/commits/7bf83ef6f1dc9396039e3358d0a44824f10690a5

File: src/devices_models/devices/electric_loads.jl (hunk 9 above)
Comment (kdheepak) [reply src/devices_models/devices/electric_loads.jl:11]: I have one last point that I have to make on this. I think it is not clear what the name of this variable should be. At the call site, we might be passing in a formulation, an instance of a formulation, the settings object etc. That is another reason I think using an underscore makes the most sense.

Comment (kdheepak) [reply src/devices_models/devices/electric_loads.jl:11]: I think we can form a consensus on the use of `_` in this fashion in a future PR. I'll go ahead and merge this for now.

File: src/devices_models/devices/renewable_generation.jl [hunk 17]
```diff
@@ -1,34 +1,26 @@
...
-        binary = false,
-        expression_name = :nodal_balance_reactive,
-    )
-end
+get_variable_binary(::ActivePowerVariable, ::Type{<:PSY.RenewableGen}) = false
+
+get_variable_expression_name(::ActivePowerVariable, ::Type{<:PSY.RenewableGen}) = :nodal_balance_active
+
+get_variable_lower_bound(::ActivePowerVariable, d::PSY.RenewableGen, _) = 0.0
+get_variable_upper_bound(::ActivePowerVariable, d::PSY.RenewableGen, _) = PSY.get_active_power(d)
```
Comment (jd-lara) [inline src/devices_models/devices/renewable_generation.jl:15]:
```suggestion
get_variable_upper_bound(::ActivePowerVariable, d::PSY.RenewableGen, _) = PSY.get_max_active_power(d)
```
